### PR TITLE
external: make rgw call separate from cephfs and rbd

### PR DESCRIPTION
**Pull request by parth-gr**
if the rgw endpoint is not reachable don't block the creation of rbd and cephfs sc, just print the error instead of raising an exception

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by travisn — COMMENTED** [review]
LGTM, but I'll defer to @BlaineEXE for final review of the python.

File: deploy/examples/create-external-cluster-resources.py [hunk 1]
```diff
@@ -1349,18 +1361,21 @@ def _gen_output_map(self):
...
+            err = ""
+            self.out_map["RGW_ADMIN_OPS_USER_ACCESS_KEY"],
+            self.out_map["RGW_ADMIN_OPS_USER_SECRET_KEY"],
+            info_cap_supported, err = self.create_rgw_admin_ops_user()
+            if err != "":
+                print(err)
+            if info_cap_supported:
+                err = self.validate_rgw_endpoint()
+            if err != "":
+                print(err)
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:1373]: @rkachach Is there any better way of doing propagation  of error?

>Using exceptions for error handilng is more Pythonic 
We were doing this before but the components are tightly connected we don't need to fail if one component doesn't work

File: deploy/examples/create-external-cluster-resources.py [hunk 2]
```diff
@@ -1275,7 +1288,7 @@ def validate_rgw_realm_name(self):
                 self._arg_parser.rgw_realm_name,
             ]
             try:
-                output = subprocess.check_output(cmd, stderr=subprocess.PIPE)
+                _ = subprocess.check_output(cmd, stderr=subprocess.PIPE)
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:1299]: removed

**Comment by parth-gr** [comment]
Currently, the script is tightly coupled with cephfs and rbd, and rgw
if  `propagate the errors` to the main function makes it more weird and complicated.

Talking about the original problem it needs to have some fix in the ocs-operator too.

Some concerns are:
1) For making it work we should make the main call to 2 separate functions one for rgw second cephfs and rbd(which require a lot of refactoring).
2) If we print the errors with the output we get output that is not directly molded to a json file.

**Comment by travisn** [comment]
> Some concerns are:
> 
> 1. For making it work we should make the main call to 2 separate functions one for rgw second cephfs and rbd(which require a lot of refactoring).

Can you expand on why a refactor would be needed? As long as we can surface all the errors, maybe we don't need a refactor?

> 2. If we print the errors with the output we get output that is not directly molded to a json file.

Usually when the output is collected for a json, it would be collected from stdout. As long as the errors are written to stderr, it wouldn't be included in the json.

**Comment by parth-gr** [comment]
This is how the output is looking if we print errors,

```
[ceph: root@rhcs-2-node-1 /]# python3 a.py --rbd-data-pool-name rbd --rgw-endpoint 10.1.115.118:8080
failed to execute command ['radosgw-admin', 'user', 'create', '--uid', 'rgw-admin-ops-user', '--display-name', 'Rook RGW Admin Ops user', '--caps', 'buckets=*;users=*;usage=read;metadata=read;zone=read', '--rgw-realm', '']. Output: b''. Code: 5. Error: b"2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 rgw_init_ioctx ERROR: librados::Rados::pool_create returned (34) Numerical result out of range (this can be due to a pool or placement group misconfiguration, e.g. pg_num < pgp_num or mon_max_pg_per_osd exceeded)\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed reading data (obj=zone.rgw.log:bucket.sync-source-hints.), r=-34\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to update sources index for bucket=:[]) r=-34\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to initialize bucket sync policy handler: get_bucket_sync_hints() on bucket=-- returned r=-34\n2023-03-28T14:51:07.223+0000 7f7ddedd1500 -1 ERROR: could not initialize zone policy handler for zone=zone\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to start notify service ((34) Numerical result out of range\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to init services (ret=(34) Numerical result out of range)\ncouldn't init storage provider\n"
[{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "rhcs-2-node-1=10.1.115.118:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "15b49e52-5bf5-11ed-a2e7-0050568fd833", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQD+SpFj/XXhGBAA3Ne52aWEFmDp33ngDyDNIA=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "10.1.115.118", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQDlT+JjhGNxIxAAriypBwv7UCMauS4S99+5lA=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQDlT+JjOVrWIxAAyVIl5R3AHoXtpg0v87RuRg=="}}, {"name": "rook-csi-cephfs-provisioner", "kind": "Secret", "data": {"adminID": "csi-cephfs-provisioner", "adminKey": "AQDlT+JjkGQVJRAAHRQ/5Pim9D4j4IbUIUg5tQ=="}}, {"name": "rook-csi-cephfs-node", "kind": "Secret", "data": {"adminID": "csi-cephfs-node", "adminKey": "AQDlT+JjA5lKJBAAccZ1BF2+OVW/yse598pFBg=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "https://10.1.115.118:8443/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "rbd", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}, {"name": "cephfs", "kind": "StorageClass", "data": {"fsName": "cephfs", "pool": "cephfs.cephfs.data", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-cephfs-node"}}]

```

**Comment by travisn** [comment]
That output looks acceptable as long as the error is going to stderr. This allows them to pipe the results to different files. For example, this should pipe the json to its own file, and the error to a separate file.
```
$ python3 a.py --rbd-data-pool-name rbd --rgw-endpoint 10.1.115.118:8080 > output.json 2> error.txt
```

**Comment by parth-gr** [comment]
> This is how the output is looking if we print errors,
> 
> ```
> [ceph: root@rhcs-2-node-1 /]# python3 a.py --rbd-data-pool-name rbd --rgw-endpoint 10.1.115.118:8080
> failed to execute command ['radosgw-admin', 'user', 'create', '--uid', 'rgw-admin-ops-user', '--display-name', 'Rook RGW Admin Ops user', '--caps', 'buckets=*;users=*;usage=read;metadata=read;zone=read', '--rgw-realm', '']. Output: b''. Code: 5. Error: b"2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 rgw_init_ioctx ERROR: librados::Rados::pool_create returned (34) Numerical result out of range (this can be due to a pool or placement group misconfiguration, e.g. pg_num < pgp_num or mon_max_pg_per_osd exceeded)\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed reading data (obj=zone.rgw.log:bucket.sync-source-hints.), r=-34\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to update sources index for bucket=:[]) r=-34\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to initialize bucket sync policy handler: get_bucket_sync_hints() on bucket=-- returned r=-34\n2023-03-28T14:51:07.223+0000 7f7ddedd1500 -1 ERROR: could not initialize zone policy handler for zone=zone\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to start notify service ((34) Numerical result out of range\n2023-03-28T14:51:07.223+0000 7f7ddedd1500  0 ERROR: failed to init services (ret=(34) Numerical result out of range)\ncouldn't init storage provider\n"
> [{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "rhcs-2-node-1=10.1.115.118:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "15b49e52-5bf5-11ed-a2e7-0050568fd833", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQD+SpFj/XXhGBAA3Ne52aWEFmDp33ngDyDNIA=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "10.1.115.118", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQDlT+JjhGNxIxAAriypBwv7UCMauS4S99+5lA=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQDlT+JjOVrWIxAAyVIl5R3AHoXtpg0v87RuRg=="}}, {"name": "rook-csi-cephfs-provisioner", "kind": "Secret", "data": {"adminID": "csi-cephfs-provisioner", "adminKey": "AQDlT+JjkGQVJRAAHRQ/5Pim9D4j4IbUIUg5tQ=="}}, {"name": "rook-csi-cephfs-node", "kind": "Secret", "data": {"adminID": "csi-cephfs-node", "adminKey": "AQDlT+JjA5lKJBAAccZ1BF2+OVW/yse598pFBg=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "https://10.1.115.118:8443/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "rbd", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}, {"name": "cephfs", "kind": "StorageClass", "data": {"fsName": "cephfs", "pool": "cephfs.cephfs.data", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-cephfs-node"}}]
> ```



> > output.json 2> error.txt

Yes you are right,

But it is not stder, its just printing of the error.
Which will be a common output

**Comment by parth-gr** [comment]
I found a solution python has 
`sys.stderr.write` so instead of `print` I will use this,

And it will separate the output

File: .github/workflows/canary-integration-test.yml [hunk 3]
```diff
@@ -167,8 +167,7 @@ jobs:
           timeout 15 sh -c "until kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool --rgw-endpoint $rgw_endpoint; do sleep 1 && echo 'waiting for the rgw endpoint to be validated'; done"
           # pass the invalid rgw-endpoint of different ceph cluster
           if output=$(timeout 15 sh -c "until kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool --rgw-endpoint 10.108.96.128:80; do sleep 1 && echo 'waiting for the rgw endpoint to be validated'; done"); then
-            echo "unexpectedly succeeded after passing the wrong rgw-endpoint: $output"
-            exit 1
+            echo "succeeded script with stderr error after passing the wrong rgw-endpoint: $output"
```
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:170]: What does it mean to succeed if there was an error? Perhaps we should say that the script completed with errors.

Comment (travisn) [reply .github/workflows/canary-integration-test.yml:170]: If they fix the problem, they will be able to re-run the script, right? There is nothing blocking them from running it again?

Comment (parth-gr) [reply .github/workflows/canary-integration-test.yml]: Yes, nothing will block them, from re-running.

It is just that they still use the output

File: .github/workflows/canary-integration-test.yml [hunk 4]
```diff
@@ -167,8 +167,7 @@ jobs:
           timeout 15 sh -c "until kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool --rgw-endpoint $rgw_endpoint; do sleep 1 && echo 'waiting for the rgw endpoint to be validated'; done"
           # pass the invalid rgw-endpoint of different ceph cluster
           if output=$(timeout 15 sh -c "until kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool --rgw-endpoint 10.108.96.128:80; do sleep 1 && echo 'waiting for the rgw endpoint to be validated'; done"); then
-            echo "unexpectedly succeeded after passing the wrong rgw-endpoint: $output"
-            exit 1
```
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:171]: Does the test not need to fail in this condition?

Comment (parth-gr) [reply .github/workflows/canary-integration-test.yml:171]: Changed the sentence too

`echo "script run completed with stderr error after passing the wrong rgw-endpoint: $output"`

So it should have failed before, but from now as the script can run with and without errors no need to fail in any case.

**Pull request merged**